### ISSUE TITLE
Some questions about the code.

### ISSUE THREAD

**Issue by q-rong** [hobincar/RecNet#13]
Hi,

In `RecNet`>>`models`>>`caption_generator. py`>>`def forward_local_reconstructor()`,

Why do you use `decoder_hiddens = build_hiddens(decoder_hiddens, caption_masks)` to calculate `decoder_hiddens`?      

Instead of using `decoder_hiddens` directly ?

What can you change about `decoder_hiddens` by using `caption_masks`?

```
    def forward_local_reconstructor(self, batch_size, decoder_hiddens, caption_masks):
        def build_hiddens(_decoder_hiddens, _caption_masks):
            _decoder_hiddens = _decoder_hiddens.permute(2, 0, 1, 3)
            _decoder_hiddens = _decoder_hiddens.view(
                _decoder_hiddens.size(0),
                _decoder_hiddens.size(1),
                _decoder_hiddens.size(2) * _decoder_hiddens.size(3)) # B, T, H
            _caption_masks = _caption_masks.transpose(0, 1).unsqueeze(2).expand_as(_decoder_hiddens).type_as(_decoder_hiddens)
            _decoder_hiddens = _caption_masks * _decoder_hiddens
            return _decoder_hiddens
        decoder_hiddens = build_hiddens(decoder_hiddens, caption_masks)
```

**Comment by hobincar** [comment]
Hi. Thank you for reporting this bug.

I thought that simply zeroing out the hidden states after "EOS" token will make the corresponding attention weights to be also zero. However, it will not gonna work.

Therefore, I modified the code so that the corresponding attention weights will be explicitly zero. Please check and comment if I got wrong again.

**Comment by q-rong** [comment]
Hi,

I am very excited to see your new revised code. But I have a new question, and I need your help.

It's hard for me to understand why you want to add `caption_masks`when calculating`weights`in the `local_reconstructor`stage.
   ```
 def forward(self, hidden, feats, masks=None):
        Wh = self.W(hidden)
        Uv = self.U(feats)
        Wh = Wh.unsqueeze(1).expand_as(Uv)
        energies = self.w(torch.tanh(Wh + Uv + self.b))
        if masks is not None:
            energies = energies.squeeze(2)
            energies[~masks] = -float('inf')
            energies = energies.unsqueeze(2)
        weights = F.softmax(energies, dim=1)

        weighted_feats = feats * weights
        attn_feats = weighted_feats.sum(dim=1)
        return attn_feats, weights
```
What I understand is that when calculating `weights`we only need `decoder_hiddens`and `hidden`, not `caption_masks`.

What's `caption_masks`, please?I didn't get it.

Is `caption_masks`a word generated by a decoder?

Looking forward to your reply, thank you.

**Comment by hobincar** [comment]
The `caption_masks` is used for masking meaningless hidden states that are generated from the decoder after it met `<EOS>` token because I padded a sentence with '<PAD>' token after the `<EOS>` token. For example, a sentence "a man is riding a bike" becomes

```
a man is riding a bike <EOS> <PAD> <PAD> ... <PAD>
```

As the number of words in each sentence is different, we cannot get rid of those meaningless hidden states by hard-coding like "discarding all hidden states after the 6th word". Therefore, the number words in each sentence should be considered to only take into account those meaningful hidden states generated from the decoder after it got valid tokens (e.g. `a`, `man`, `is`, `riding`, `bike`).

**Comment by q-rong** [comment]
Hi,

I have a question that really confuses me, and I think it might be a bug.

In the validation phase, you set `teacher_forcing_ratio=0.`. We no longer need to get `caption`from a dataset such as MSVD. So why do you use `captions` to get `captions_mask` and then `captions_mask` to get `feats_recon`?

```
utils.py >> def test():

    for b, batch in enumerate(val_iter, 1):
        _, feats, captions = parse_batch(batch)
        output, feats_recon = model(feats, captions)
```
```
caption_generator.py >> def forward():

        caption_masks = captions != self.vocab.word2idx['<PAD>']
        caption_masks = caption_masks.cuda()
                            ...        
        if self.reconstructor is not None:
               feats_recon = self.forward_reconstructor(batch_size, decoder_hiddens
                                                                         , caption_masks)
```
```
caption_generator.py >> def forward_local_reconstructor():

          for t in range(self.decoder.feat_len):
               _, hidden = self.reconstructor(decoder_hiddens, hidden, caption_masks)
               feats_recons[t] = hidden[0] if self.decoder.rnn_type == 'LSTM' else hidden
```



The specific code is as follows:


 ```
caption_generator. py >> def forward():

   def forward(self, feats, captions=None, teacher_forcing_ratio=0.):
        batch_size = feats.size(0)
        vocab_size = self.decoder.output_size
        caption_masks = captions != self.vocab.word2idx['<PAD>']
        caption_masks = caption_masks.cuda()

        hidden = self.get_rnn_init_hidden(batch_size, self.decoder.num_layers, self.decoder.num_directions,
                                          self.decoder.hidden_size)
        outputs, decoder_hiddens = self.forward_decoder(batch_size, vocab_size, hidden, feats, captions,
                                                        teacher_forcing_ratio)
        feats_recon = None
        if self.reconstructor is not None:
            feats_recon = self.forward_reconstructor(batch_size, decoder_hiddens, caption_masks)
        return outputs, feats_recon
```

**Comment by hobincar** [comment]
Yes, we don't need the ground-truth captions as we set the `teacher_forcing_ratio=0` in the validation phase. I should have generated the `caption_masks` from the `outputs` of the decoder instead of the ground-truth `captions`. I'll modify it soon.

However, this would not have affected the training of the model, since it's an error in calculating the validation loss of the reconstructor. We just have seen a wrong validation error of the reconstructor.

**Comment by q-rong** [comment]
Hi,

I'm excited to see you update the code. I have two problems that have been bothering me, which I really can't understand.

(1) I don't understand why `hiddens` is defined as follows:
 ```
caption_generator.py >> def forward_decoder():

hiddens = Variable(
            torch.zeros_like(hidden[0] if self.decoder.rnn_type == 'LSTM' else hidden)
                .unsqueeze(0)
                .repeat(self.max_caption_len + 2, 1, 1, 1))
``` 
I wonder why `hiddens` is not defined as follows. In short, `hiddens` and `outputs` are defined in the same way.
 ```
hiddens = Variable(torch.zeros(self.max_caption_len + 2, batch_size, vocab_size))
``` 
(2) Meanwhile, I don't understand why `hidden[0] if self.decoder.rnn_type == 'LSTM' else hidden`, why `hidden[0]` instead of `self.decoder.rnn_type == 'LSTM'`?


The specific code is as follows:
```
caption_generator.py >> def forward_decoder():

    def forward_decoder(self, batch_size, vocab_size, hidden, feats, captions, teacher_forcing_ratio):
        outputs = Variable(torch.zeros(self.max_caption_len + 2, batch_size, vocab_size)).cuda()
        hiddens = Variable(
            torch.zeros_like(hidden[0] if self.decoder.rnn_type == 'LSTM' else hidden)
                .unsqueeze(0)
                .repeat(self.max_caption_len + 2, 1, 1, 1))

        output = Variable(torch.cuda.LongTensor(1, batch_size).fill_(self.vocab.word2idx['<SOS>']))
        for t in range(1, self.max_caption_len + 2):
            output, hidden, attn_weights = self.decoder(output.view(1, -1), hidden, feats)
            outputs[t] = output
            hiddens[t] = hidden[0] if self.decoder.rnn_type == 'LSTM' else hidden
            is_teacher = random.random() < teacher_forcing_ratio
            top1 = output.data.max(1)[1]
            output = Variable(captions.data[t] if is_teacher else top1).cuda()
        return outputs, hiddens
```

**Comment by hobincar** [comment]
(1) I saved decoder hidden states into the variable `hiddens` where the shape of a decoder hidden state is `(D=1, B=200, H=512)`. Therefore, the shape of the variable `hiddens` should be `(self.max_caption_len + 2, D, B, H)`. The shape of the variable `outputs` should be `(self.max_caption_len + 2, B, V=13380)`. BTW, the code was hard to read. I'll modify this to be more readable.

(2) When we use LSTM as a decoder, the variable `hidden` will be a tuple of two tensors `(HIDDEN_STATE, CONTEXT_STATE)`. When we use GRU as a decoder, the variable `hidden` will be a tensor `HIDDEN_STATE`. However, we only need the `HIDDEN_STATE` for our reconstructor's input. That's why I divide the cases into LSTM / GRU.

**Comment by q-rong** [comment]
Thank you for your patient reply. Your explanation is very clear.

(1) I'm wondering why you're calculating `loss`, why is `reg_lambda = 0`?
```
utils.py >> train() / test():
        cross_entropy_loss = F.nll_loss(output[1:].view(-1, vocab.n_vocabs),
                                        captions[1:].contiguous().view(-1),
                                        ignore_index=PAD_idx)
        entropy_loss = losses.entropy_loss(output[1:], ignore_mask=(captions[1:] == PAD_idx))
        loss = cross_entropy_loss + reg_lambda * entropy_loss
```
(2) During the training of the model, I observed that `train loss` presented a downward trend, while `val loss` presented a downward trend and then an upward trend. Even the final `val loss` was far more than the `val loss` of `Epoch 1`.

Is this a normal behavior? Why does `train loss` decline while `val loss` rise? Is this overfitting?

**Comment by hobincar** [comment]
(1) The `reg_lambda` denotes the loss weight to entropy loss, which forces our model to predict the next word with high probability by minimizing entropy. It is a trivial technique to enhance performance. However, what I wanted was reproducing the performance of the original paper, I set it to zero for now. You can increase it and see the changes in performance.

(2) I also noticed that kind of increase in validation loss. I'm also curious about that point but just ignored it for now as the captioning performances (BLEU@4, CIDEr-D, etc.) are increasing. I'm sorry that I cannot give you some satisfactory answers.

**Comment by q-rong** [comment]
> 
> 
> (1) I saved decoder hidden states into the variable `hiddens` where the shape of a decoder hidden state is `(D=1, B=200, H=512)`. Therefore, the shape of the variable `hiddens` should be `(self.max_caption_len + 2, D, B, H)`. The shape of the variable `outputs` should be `(self.max_caption_len + 2, D, B, V=512)`. BTW, the code was hard to read. I'll modify this to be more readable.
> 
> (2) When we use LSTM as a decoder, the variable `hidden` will be a tuple of two tensors `(HIDDEN_STATE, CONTEXT_STATE)`. When we use GRU as a decoder, the variable `hidden` will be a tensor `HIDDEN_STATE`. However, we only need the `HIDDEN_STATE` for our reconstructor's input. That's why I divide the cases into LSTM / GRU.

Hi,

(1) I notice you say  `The shape of the variable 'outputs' should be '(self.max_caption_len + 2, D, B, V=512)'`, but in your code it's `outputs = Variable(torch.zeros(self.max_caption_len + 2, batch_size, vocab_size=13380)).cuda()`. 

I found that one parameter was missing. Is this an error?

(2) I don't understand why the variable `hiddens` has two tensors`(HIDDEN_STATE, CONTEXT_STATE)`.
What is `CONTEXT_STATE` and what does it do? What does `HIDDEN_STATE` have to do with `CONTEXT_STATE`?

(3) Why do you want to `bottleneck_size` set to 256, rather than the other values, such as `512` or `1536` and so on.  Reduce `self.feat_size=1536` and `self.hidden_size=512` to 256. Won't there be data loss?
```
class TemporalAttention(nn.Module):
    def __init__(self, hidden_size, feat_size, bottleneck_size):
        super(TemporalAttention, self).__init__()
        self.hidden_size = hidden_size
        self.feat_size = feat_size
        self.bottleneck_size = bottleneck_size
        self.W = nn.Linear(self.hidden_size, self.bottleneck_size, bias=False)
        self.U = nn.Linear(self.feat_size, self.bottleneck_size, bias=False)
        self.b = nn.Parameter(torch.ones(self.bottleneck_size), requires_grad=True)
        self.w = nn.Linear(self.bottleneck_size, 1, bias=False)
```
